### PR TITLE
feature: Addition of memory snippets

### PR DESCRIPTION
**Pull request by esolithe**
A variant of the ideas of World Info and Vector based Long Term Memory.  While within Lite we cannot expect vector storage to exist - at least not in its traditional form of embeddings any time soon, it is possible to implement the searching algorithms they use to extract information from the entire game text which are relevant.  These are based on similarity, and after testing a few I settled on Tf-Idf - which is explained more in the links I provided in the comments within the new code.

The general basis looks for frequency of distinctive words, the snippet length and a couple of factors like that - to try and provide a better search while still remaining performant.  While this is still a fairly rudimentary implementation / integration within Lite, it runs entirely browser side and seems to do alright with finding relevant matches.

With the implementation, there is the addition of a new menu under context which allows enabling the new option, modifying the minimum similarity required to find matches and how many matches to include (similar to Top N).  The matches are included with the world info as an authors note formatted block with the different snippets.

While I have tried to generally match formatting / style with the existing code, as this is my first PR for Lite apologies in advance if there's anything out of place!

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -18852,4 +18929,495 @@
...
+		/**
+		 * Max length per chunk
+		 */
+		_chunkSplitSize
+		/**
+		 * Overlap per chunk
+		 */
+		_chunkOverlap
+
+		constructor(amountToTrimForContext, maxLengthOfSearchString = 400, chunkSplitSize = 600, chunkOverlap = 200) {
```
Comment (BBC-Esq) [inline index.html:19001]: This can sometimes make a difference if we want to allow a user to change it?

File: index.html [hunk 2]
```diff
@@ -18852,4 +18929,495 @@
...
+		 * Escapes a string to a regex supported format
+		 */
+		_escapeRegExp(stringToEscape) {
+			return stringToEscape.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
+		}
+
+		/**
+		 * Gets the last response from the AI and yourself
+		 */
+		getLastResponse() {
```
Comment (BBC-Esq) [inline index.html:19113]: Maybe add here to remove all specified characters like punctuation marks?  Not sure if this is the right portion...

```this._lastResponseCache = this._lastResponseCache.replace(/[.,\/#!$%\^&\*;:{}=\-_`~()\[\]]/g, "");```

**Comment by BBC-Esq** [comment]
Hope I did this right...

File: index.html (hunk 2 above)
Comment (esolithe) [reply index.html:19113]: I do not think this is needed.  Different searchers in future may require those characters - so I would say leave them.  This is negated by the Minisearch lib anyway, as it splits by words and gets rid of special characters anyway as far as I understand it.

File: index.html (hunk 1 above)
Comment (esolithe) [reply index.html]: Done

**Comment by LostRuins** [comment]
I have yet to try the improved version, which I will review once you two are done with it

**Review by BBC-Esq — APPROVED** [review]
Trust your judgment unless there's a specific bug, which I can help troubleshoot.

**Comment by esolithe** [comment]
> Was there something to review or you mentioned that you fixed the issue?

As discussed in the chat on Discord, the requested settings should now be UI options.  Henky is giving it a quick go I believe, and the library used can be found here: https://github.com/lucaong/minisearch 

And the minified build in particular here: https://www.jsdelivr.com/package/npm/minisearch (https://cdn.jsdelivr.net/npm/minisearch@7.1.0/dist/umd/index.min.js)

**Comment by esolithe** [comment]
This should now be ready for another review - hopefully a bit more reliable than last time.  As mentioned on discord, please note the following:

- Seems to take about 10-15 seconds for a 5mb file (which given the max file size of lite at the moment should be fine I think).  There are some areas I saw which might be possible to optimise, but they are mostly at a library level.
- This does not replace / work too well at smaller context sizes - while it does help when you're using 8K and up, due to the amount of text it extracts due to the chunking it comes with some caveats - but it is seems to be a fair bit more stable than it used to be.
- I was going to implement caching of the index, but given it seems to run in 10 second or so for that amount of text, I'll imagine it should be fine for now - maybe a possible future optimisation.

**Comment by LostRuins** [comment]
I have a bunch of changes I wanna make, but I think i'll merge the current version and apply them on top of it so we have something to look back to for reference in future.

**Pull request merged**